### PR TITLE
Audio Recording from godot

### PR DESCRIPTION
**Pull request by Tugsav**
This pull request implements an Audio bus effect, that outputs the audio from the bus into a raw audio file.

I implemented this since i needed it myself but figured i would put it up as a PR in case it could be useful for others.

Potential use case examples:
- Record audio for in-game cut-scene
- Allowing the user to save specific audio during run-time, for an example if you have a recorder in-game.

[edit]
As per the suggestions, the functionality has been changed in the following ways
The audio effect now records into a buffer that a user can get an AudioStreamSample object from.
AudioStreamSample formats supported are 8-bit and 16-bit PCM, as well as IMA_ADPCM
Additionally a function has been added to AudioStreamSample that allows the user to save it as a WAV file.
Supported formats for saving as WAV are 8-bit and 16-bit PCM

### CODE REVIEW

File: servers/audio/effects/audio_effect_record.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/*************************************************************************/
+/*  audio_effect_record.cpp                                             */
```
Comment (Noshyaar) [inline servers/audio/effects/audio_effect_record.cpp]: Please keep the `*/` in line on both files.

Thanks for contributing! Very nice feature. :)

Comment (Tugsav) [reply servers/audio/effects/audio_effect_record.cpp]: Fixed. Thanks for spotting it quickly

**Comment by Zylann** [comment]
Aaah I thought it was for recording with a mic ^^"

**Comment by groud** [comment]
How do you change the output format ?

**Comment by Tugsav** [comment]
The output format is raw audio (no format really). In order to play it with a media player you have to encode it to a real format.
As an example i used ffmpeg to make a wav file:
`ffmpeg -f f32le -ar 44.1k -ac 2 -i testaudio.raw testconverted.wav`

**Comment by groud** [comment]
I like the idea, but I wonder if how it is useful when building a game.
I think it would be better making this a plugin, as I see few realistic use cases.

But maybe you have a good counter-example. ^^

**Comment by Tugsav** [comment]
The original intention of this were to use it for pre-rendered cutscenes created in the godot engine (using the --fixed-fps and --disable-render-loop flags and using a GD-script to save frames).
As for whether to use pre-rendered cutscenes or not (or a mix) the opinions are many ^^'

**Comment by reduz** [comment]
Memory allocation and IO in audio threads is forbidden, you should probably use a ringbuffer for this and write in another thread.

Also, we should expose AudioEffect via GDNative, so you can do this without needing to modify core godot.

**Comment by Zylann** [comment]
Being able to write AudioEffects using GDNative would be awesome :D

**Comment by Tugsav** [comment]
I have moved the IO to a separate thread and instead buffered the Audio as requested. Also the recording audio effect now creates a header in the file so that we get valid (though completely uncompressed) .wav files.

**Comment by Tugsav** [comment]
I have rebased the PR to adhere to the guidelines. Also the audio is now saved as uncompressed .wav files instead of raw audio.

**Comment by reduz** [comment]
@SaracenOne is working on something similar that also includes audio input. Holding off on this one to see how both mix, if they do.

**Comment by SaracenOne** [comment]
@reduz It should be fine I think. As per what we discussed, my microphone code is only interested in trying to get microphone input from the audio driver into an audio stream, and then ultimately into the mixer where we can do things with it.

**Comment by reduz** [comment]
After discussion, probably a more desired functionality would be to just record to an AudioStreamSample, and eventually add a function to it to save to a wav file. This way we can use it together with the new option to record from microphone

**Comment by Tugsav** [comment]
@reduz PR has been updated in an attempt to implement it like the desired functionality. PR description has been updated as well.

**Comment by akien-mga** [comment]
There seems to be a build issue on Android: https://travis-ci.org/godotengine/godot/jobs/407176455 (you might need to press the "Raw log" button if it doesn't show up)

**Comment by Tugsav** [comment]
I will look into those that build issue. It seems to be linked to the fact that i made`ResourceImporterWAV::_compress_ima_adpcm` into a static method.
Not sure if i should fix the commented code or just squash both commits into a single one.

**Comment by akien-mga** [comment]
>  Not sure if i should fix the commented code or just squash both commits into a single one.

That's up to you, but I think it makes sense to have two commits, once for implementing the audio recording feature, and one for implement the save to WAV.

**Comment by Tugsav** [comment]
Recommended changes have been applied, and the Android build issue resolved

**Comment by akien-mga** [comment]
Awesome, thanks!

**Pull request merged**